fission-server/src/models/account.rs

```diff
@@ -37,59 +44,155 @@ pub struct Account {
     }
 
-    pub async fn find_by_username_and_did(
-        mut conn: Conn<'_>,
+    /// Find a Fission Account by username, validate that the UCAN has permission to access it
+    pub async fn find_by_username(
+        conn: Arc<Mutex<Conn<'_>>>,
+        ucan: ucan::Ucan,
         username: String,
-        did: String,
     ) -> Result<Self, diesel::result::Error> {
-        accounts::dsl::accounts
+        let mut conn = conn.lock().await;
+
+        let account = accounts::dsl::accounts
             .filter(accounts::username.eq(username))
-            .filter(accounts::did.eq(did))
-            .first(&mut conn)
-            .await
+            .first::<Account>(&mut conn)
+            .await;
+
+        // FIXME this should actually validate that the UCAN has access, not just that the DID matches
+        if let Ok(account) = account {
+            if *ucan.issuer().to_string() == account.did {
+                return Ok(account);
+            }
+        }
+
+        Err(diesel::result::Error::NotFound)
     }
 
+    /// Update the controlling DID of a Fission Account
+    //
+    // Unlike the update_volume_cid method below, this could be done with one SQL query, but
+    // for consistency, we're using two; fetch the account and then update it, as separate operations.
+    //
+    // There's probably an elegant way to do that chaining with Rust and Diesel, but I don't know how
+    // to do it and (1) it's not a huge deal performance-wise, and (2) it would be a lot more complex.
     pub async fn update_did(
-        mut conn: Conn<'_>,
-        username: String,
-        old_did: String,
+        &self,
+        conn: Arc<Mutex<Conn<'_>>>,
         new_did: String,
     ) -> Result<Self, diesel::result::Error> {
-        diesel::update(
-            accounts::dsl::accounts
-                .filter(accounts::username.eq(username))
-                .filter(accounts::did.eq(old_did)),
-        )
-        .set(accounts::did.eq(new_did))
-        .get_result(&mut conn)
-        .await
+        // FIXME this needs to account for delegation and check that the correct
+        // capabilities have been delegated. Currently we only support using the root did.
+        let mut conn = conn.lock().await;
+        diesel::update(accounts::dsl::accounts)
+            .filter(accounts::id.eq(self.id))
+            .set(accounts::did.eq(new_did))
+            .get_result(&mut conn)
+            .await
+    }
+
+    /// Get the volume associated with the user's account.
+    //
+    // Note: this doesn't use a join, but rather a separate query to the volumes table.
+    // Possibly not ideal, but it's simple and works.
+    pub async fn get_volume(
+        &self,
+        conn: Arc<Mutex<Conn<'_>>>,
+        // ucan: ucan::Ucan,
+        // nb not including the ucan here, because in order to get the account,
+        // we've already validated the UCAN. HOWEVER, we should probably validate
+        // that the UCAN has access to the volume here.
```

Comment:
I don't think this is correct. I think everyone should be able to access the volume of any user, because you'll want to load their public files.